### PR TITLE
Add CMS AsyncAPI service

### PR DESCRIPTION
**Pull request by michal-hudy**
**Description**

Changes proposed in this pull request:

- Added CMS AsyncAPI service

**Related issue(s)**

#4960

### CODE REVIEW

File: resources/cms/charts/cms-asyncapi-service/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        type: RollingUpdate
+        rollingUpdate:
+          maxUnavailable: 0
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{ .Values.global.containerRegistry.path }}/{{ .Values.global.cmsAsyncAPIService.dir }}{{ .Values.image.name }}:{{ .Values.global.cmsAsyncAPIService.version }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+            - name: APP_VERBOSE
+              value: "true"
```
Comment (derberg) [inline resources/cms/charts/cms-asyncapi-service/templates/deployment.yaml]: move to values.yaml

File: resources/cms/charts/cms-asyncapi-service/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            - name: APP_SERVICE_PORT
+              value: "{{ .Values.service.internalPort }}"
+            - name: APP_SERVICE_HOST
+              value: "0.0.0.0"
```
Comment (derberg) [inline resources/cms/charts/cms-asyncapi-service/templates/deployment.yaml]: move to values.yaml

Comment (michal-hudy) [reply resources/cms/charts/cms-asyncapi-service/templates/deployment.yaml]: I disagree, this value shouldn't be exposed in `values.yaml`

File: components/cms-services/pkg/runtime/signal/signal.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+package signal
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+// SetupChannel registered for SIGTERM and Interrupt. A stop channel is returned
```
Comment (kazydek) [inline components/cms-services/pkg/runtime/signal/signal.go]:
```suggestion
// SetupChannel is registered for SIGTERM and Interrupt. A stop channel is returned
```

File: components/cms-services/pkg/runtime/signal/signal.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// which is closed on one of these signals. If a second signal is caught, the program
```
Comment (kazydek) [inline components/cms-services/pkg/runtime/signal/signal.go]:
```suggestion
// and closed on one of these signals. If the second signal is caught, the program
```

File: components/cms-services/pkg/runtime/signal/signal.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		<-c
+		close(stop)
+		<-c
+		os.Exit(1) // second signal. Exit directly.
+	}()
+
+	return stop
+}
+
+// CancelOnInterrupt calls cancel function when os.Interrupt or SIGTERM is received
```
Comment (kazydek) [inline components/cms-services/pkg/runtime/signal/signal.go]:
```suggestion
// CancelOnInterrupt calls the cancel function when os.Interrupt or SIGTERM is received.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/kyma-project/kyma/components/cms-services/pkg/runtime/endpoint"
+	"github.com/kyma-project/kyma/components/cms-services/pkg/runtime/service"
+
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+)
+
+// Convert is a functional mutation handler that converts AsyncAPI specification
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go]:
```suggestion
// Convert is a functional mutation handler that converts the AsyncAPI specification.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+type Convert func(reader io.Reader, writer io.Writer) error
+
+var _ endpoint.Mutator = Convert(nil)
+
+// Mutate convert AsyncAPI spec from version 1.* to version 2.0.0-rc1
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go]:
```suggestion
// Mutate converts the AsyncAPI specification from version 1.* to version 2.0.0-rc1.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	if err := c(reader, writer); err != nil {
+		return nil, err
+	}
+	if err := writer.Flush(); err != nil {
+		return nil, err
+	}
+	return b.Bytes(), nil
+}
+
+// AddConversion registers endpoint in service
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/conversion_endpoint.go]:
```suggestion
// AddConversion registers the endpoint in a service.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+func AddValidation(srv service.Service) error {
+	validator := Validate(parser.Parse)
+
+	srv.Register(endpoint.NewValidation("v1/validate", validator))
+	return nil
+}
+
+var _ endpoint.Validator = Validate(nil)
+
+// Validate is a functional validation handler that checks AsyncAPI schema
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go]:
```suggestion
// Validate is a functional validation handler that checks the AsyncAPI schema.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+type Validate func(yamlOrJSONDocument []byte, circularReferences bool) (json.RawMessage, *errs.ParserError)
+
+// Validate checks AsyncAPI spec against 2.0.0-rc1 schema
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go]:
```suggestion
// Validate checks the AsyncAPI specification against the 2.0.0-rc1 schema.
```

File: components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+func (v Validate) Validate(ctx context.Context, reader io.Reader, parameters string) error {
+	document, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return errors.Wrapf(err, "while reading content")
```
Comment (kazydek) [inline components/cms-services/pkg/endpoint/asyncapi/v1/validation_endpoint.go]:
```suggestion
		return errors.Wrapf(err, "while reading the content")
```

File: components/cms-services/cmd/asyncapi/openapi.yaml [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        "422":
+          $ref: "#/components/responses/unprocessableEntity"
+        "500":
+          $ref: "#/components/responses/internalServerErrorResponse"
+servers:
+  - url: http://localhost:3000/v1
+components:
+  responses:
+    badRequestResponse:
+      description: Bad request response
```
Comment (kazydek) [inline components/cms-services/cmd/asyncapi/openapi.yaml]:
```suggestion
      description: Bad request
```

File: components/cms-services/cmd/asyncapi/openapi.yaml [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      content:
+        text/plain:
+          schema:
+            type: string
+            example: Invalid format
+    methodNotAllowedResponse:
+      description: Bad request response
```
Comment (kazydek) [inline components/cms-services/cmd/asyncapi/openapi.yaml]:
```suggestion
      description: Bad request
```

File: components/cms-services/cmd/asyncapi/openapi.yaml [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        text/plain:
+          schema:
+            type: string
+            example: Internal server error
+    successResponse:
+      description: Success response
+    notModified:
+      description: File content not modified
+    successConversionResponse:
+      description: Success response with new file content
```
Comment (kazydek) [inline components/cms-services/cmd/asyncapi/openapi.yaml]:
```suggestion
      description: Success response with the new file content
```

**Pull request merged**